NEw ST_Angle function, with test and doc

Remi-C opened the pull request:
Hey,
new function with 2 behaviours,
ST_Angle(P1,P2,P3) returns the angle of P1P2P3
ST_ANgle(P1,P2,P3,P4) returns the angle of 2 vectors P1P2,P3P4

For an easy visual test, use the query in the doc to generate a few dozen angles, and display in qgis.
Cheers

strk commented on postgis/lwgeom_functions_basic.c:
```diff
@@ -2399,6 +2400,88 @@ Datum LWGEOM_azimuth(PG_FUNCTION_ARGS)
+			n_args = i ; 
+			if (n_args < 3)
+				lwpgerror("Empty geometry"); 
+			break; 
```
Should this be a PG_RETURN_NULL() instead ?

strk replied:
Sorry, I meant within the conditional block, not outside

strk commented on doc/reference_measure.xml:
```diff
@@ -820,6 +820,64 @@ SELECT degrees(ST_Azimuth(ST_Point(25, 45), ST_Point(75, 100))) AS degA_B,
+		<refnamediv>
+		  <refname>ST_Angle</refname>
+
+		  <refpurpose>Returns the angle in radian measured clokwise between 3 points, or the angle measured clockwise between 2 vectors.</refpurpose>
```
"clokwise" -> "clockwise" (first occurrence)

If I may suggest also keeping the "refpurpose" text small (may omit radian and clockwise, leaving it specified in the "description")

strk commented on doc/reference_measure.xml:
```diff
@@ -820,6 +820,64 @@ SELECT degrees(ST_Azimuth(ST_Point(25, 45), ST_Point(75, 100))) AS degA_B,
+			  <paramdef><type>geometry </type><parameter>point1</parameter></paramdef>
+			  <paramdef><type>geometry </type><parameter>point2</parameter></paramdef>
+			  <paramdef><type>geometry </type><parameter>point3</parameter></paramdef>
+			  <paramdef><type>geometry </type><parameter>point4</parameter></paramdef>
```
Isn't the 4th point optional ? There should be a tag or attribute to specify that in the docs.

strk commented on doc/reference_measure.xml:
```diff
@@ -820,6 +820,64 @@ SELECT degrees(ST_Azimuth(ST_Point(25, 45), ST_Point(75, 100))) AS degA_B,
+			<para> For 3 points, computes the angle measured clockwise of P1P2P3.
+			For 4 points,compute the angle measured clockwise of P1P2,P3P4.
+			Results are always positive, between 0 and 2*Pi radians.
+			Uses ST_Azimuth.
```
"Uses ST_Azimuth" isn't really true, it's using liblwgeom internal functions (on which ST_Azimuth is also based).

strk commented on postgis/lwgeom_functions_basic.c:
```diff
@@ -2399,6 +2400,88 @@ Datum LWGEOM_azimuth(PG_FUNCTION_ARGS)
+		lwpoint = lwgeom_as_lwpoint(geom_unser);
+		if ( ! lwpoint )
+		{
+			PG_FREE_IF_COPY(geom, i);
```
These memory releases should really affect all the previous geometry inputs, or only the last one would be released.

strk commented on postgis/lwgeom_functions_basic.c:
```diff
@@ -2399,6 +2400,88 @@ Datum LWGEOM_azimuth(PG_FUNCTION_ARGS)
+		if ( ! getPoint2d_p(lwpoint->point, 0, &points[i]) )
+		{
+			PG_FREE_IF_COPY(geom, i);
+			lwpgerror("Error extracting point");
```
This should really only happen when the point is empty (hint hint [ for reducing complexity ])

strk commented on postgis/lwgeom_functions_basic.c:
```diff
@@ -2399,6 +2400,88 @@ Datum LWGEOM_azimuth(PG_FUNCTION_ARGS)
+	if (n_args == 3)
+	{
+		if ( ! azimuth_pt_pt(&points[0], &points[1], &az1) )
+			PG_RETURN_NULL(); 
```
I guess the code here expects azimuth_pt_pt to call lwerror itself, better specify it with a comment, or a reader would be wondering why no error message is printed.

An lwpgerror could even be left to be even ready shall liblwgeom stop reporting errors in such a primitive way :)

strk commented on regress/lwgeom_regress_expected:
```diff
@@ -21,3 +21,11 @@ BoundingDiagonal3|SRID=4326;LINESTRING(999999986991104 999999986991104,999999986
+ST_Angle_4_pts|4.71238898038469
+ST_Angle_4_pts|0.785398163397448
+ST_Angle_3_pts|1.5707963267949
+ERROR:  Operation on mixed SRID geometries
```
The EMPTY case is missing a test

Remi-C replied:
Hey,
I removed the empty test case because ST_Azimuth is failing it.
(it returns another error than the one expected)
it returns `ERROR:  getPoint2d_p: point offset out of range`

Remi-C replied:
hm never mind its ok I added the missing test, thanks

Remi-C replied:
hm never mind its ok I added the missing test, thanks

strk replied:
> I removed the empty test case because ST_Azimuth is failing it.
> (it returns another error than the one expected)
> it returns `ERROR:  getPoint2d_p: point offset out of range` 

Evidently your is_empty calls are not working.
Better fix that first.

Remi-C commented:
Hey @strk , thanks for the review!
I don't get the PG_RETURN_NULL stuff,
what do you mean exactly?
I corrected the other things you mention, know working on the memory stuff.
Cheers

strk commented:
I just meant that any "lwpgerror" call should be followed by
a PG_RETURN_NULL(), which wasn't done in the n_args < 3
check (btw, on re-reading that check is probably bogus
and the source of your problems with empty ?)

Remi-C commented:
@strk I don't get the  memory issue you mention.
`POINT2D points[4]; ` doesn't need to be freed manually I guess.
Other variables are freed in each loop, so no need to free those in case of error some loop after?
Do you mean I should free `geom_unser` and / or `lwpoint`
before each `PG_RETURN_NULL` ? 
Cheers

strk commented:
I mean the PG_FREE_IF_COPY calls you add.
They are inconsistent.
Ideally you'd have a PG_FREE_IF_COPY call for each
and every parameter, before exiting or raising an error.

Remi-C commented:
Hey @strk, here is a rewrite,
with better separation of exception and actual computing.
Exception are tried on serialized geometry.
About  `PG_FREE_IF_COPY`,
from what I understand there is no need to use it on a parameter if you havn't used `PG_GETARG_GSERIALIZED_P`.
If this is true, it should be ok as it is now (only one arg is in memory at a time, being cleaned at the end of the loop anyway)
Cheers,
Remi-C

strk commented:
Ok so I see you're detoasting/free-if-copying once for checking and again for actual using.
Could you avoid the duplication ?

strk commented on postgis/lwgeom_functions_basic.c:
```diff
@@ -2399,6 +2400,124 @@ Datum LWGEOM_azimuth(PG_FUNCTION_ARGS)
+		if (!lwpoint)
+		{ 
+			lwpgerror("Error unserializing geometry"); 
+			PG_RETURN_NULL() ;
```
Also this misses a PG_FREE_IF_COPY...

Remi-C replied:
@strk Ok as you suggested I stored the serialized pointers to keep getting
those twice.
Now `PG_FREE_IF_COPY` is always executed if exceptions or in regular
operations.

2016-02-26 14:47 GMT+01:00 Sandro Santilli <notifications@github.com>:

> In postgis/lwgeom_functions_basic.c
> <https://github.com/postgis/postgis/pull/97#discussion_r54244102>:
>
> > +			
> > +			case 3:
> > +			lwpgerror("Operation on mixed SRID geometries");
> > +			PG_RETURN_NULL();
> > +			break;
> > +		}
> > +	/* extract points */
> > +	for(i=0; i<n_args; i++)
> > +	{
> > +		geom = PG_GETARG_GSERIALIZED_P(i);
> > +		geom_unser = lwgeom_from_gserialized(geom) ;
> > +		lwpoint = lwgeom_as_lwpoint(geom_unser);
> > +		if (!lwpoint)
> > +		{
> > +			lwpgerror("Error unserializing geometry");
> > +			PG_RETURN_NULL() ;
>
> Also this misses a PG_FREE_IF_COPY...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/postgis/postgis/pull/97/files#r54244102>.
>

Remi-C commented:
Hey @strk , here the pointers are stored to avoid calling twice the functions.
Cheers

strk commented on postgis/lwgeom_functions_basic.c:
```diff
@@ -2399,6 +2400,129 @@ Datum LWGEOM_azimuth(PG_FUNCTION_ARGS)
+		/* dont do lwpoint_free(lwpoint); , this memory is needed ! */ 
+	}
+	for (j=0;j<n_args;j++)
+		PG_FREE_IF_COPY(seri_geoms[j], j);
```
It is not safe to release the serialized format after deserializing, because LWGEOM can still refer to the serialized POINTLIST.

Remi-C commented:
@strk : thanks to help me learn all of this.
Cheers,
RemiC

robe2 commented:
@Remi-C  looks like this isn't ready yet as @strk had one last note doesn't look like you addressed.  Also this needs to be rebased if to be committed for PostGIS 2.4.

robe2 commented:
@Remi-C  can you rebase and I'll test and  commit for PostGIS 2.4

Remi-C commented:
Hey Regina, @strk,
I removed the serialized memory freeing.

About my rebasing,
I think I messed something,
I rebased so my branch st_angle is updated and contains all the latest
postgis commits.
What I don't understand is why now the pull request contains also all the
postgis commits?

Shall we drop this pull request and make a new one?
Cheers,
Rémi-C

2017-08-10 12:09 GMT+02:00 Regina Obe <notifications@github.com>:

> @Remi-C <https://github.com/remi-c> can you rebase and I'll test and
> commit for PostGIS 2.4
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/postgis/postgis/pull/97#issuecomment-321509968>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEgcIoFwhfvovWvSLMFRRbGjAPDuDO-Uks5sWtbegaJpZM4Hi4Ia>
> .
>

strk commented:
Try this (assuming "origin" points to the official mirror):

  git rebase origin
  git push -f

robe2 commented:
keep in mind it's too late for this to go into 2.4.  Has to wait for 2.5

Remi-C commented:
Hey @robe2, no problem, it'll be for 2.5 with any luck.
Thanks @strk, this fixed it (after some awkward manual edits)

Minor typos
